### ISSUE TITLE
Consumer receives messages out of order

### ISSUE THREAD

**Issue by Oduig** [rabbitmq/rabbitmq-server#1775]
We are encountering an issue with RabbitMQ consumers (Spring Boot) where messages regularly arrive out of order. We created a minimal example that uses internal libraries. Instead of sharing those, I extracted the relevant pieces of code that show the problem.

Is there anything that needs to be configured to guarantee ordering in RabbitMQ? If not, what might be causing our messages to consistently arrive out of order?

## Environment

Build: Docker image `rabbitmq:3.7-alpine` with RabbitMQ version `3.7.8`
See https://github.com/docker-library/rabbitmq/blob/master/3.7/alpine/Dockerfile

Java client using Spring Boot: `spring-boot-starter-amqp:2.0.0-RELEASE`.
See https://mvnrepository.com/artifact/org.springframework.boot/spring-boot-starter-amqp/2.0.0.RELEASE

## Steps to reproduce

1. Set up a single RabbitMQ node with the settings provided by the Docker image.
2. Configure the standard `spring.rabbitmq` settings.
3. Create a queue from Java (durable, non-exclusive, no auto-delete), see `componentQueue` below.
4. Start a thread that publishes strictly ordered messages to the queue, see `MessageEmitter` below.
5. Create a RabbitListener from Java (single-threaded, max 1 consumer), see `rabbitListenerContainerFactory` below.
6. Register a message listener, see `onMessageReceived` below.
7. Observe the logs.

## Logs

The code below writes approximately `150` messages per second to a single exchange and subscribes to that exchange with a single-consumer queue. It uses atomic primitives to protect against concurrency issues. In that way, we would expect out-of-order messages to be an uncommon occurrence.

However, the logs are completely saturated with messages like these. Note that only messages out of order are logged, so when it says "arrived 13 too late, skipped 14", that means one message was out of order and we are once again reading from the head of the queue.

```
2018-11-19 10:11:19.454 ERROR 1 --- [cTaskExecutor-1] n.g.v.s.rabbitmq.tester.MessageLogger    : Received message 00053406 - SKIPPED 1 MESSAGES.
2018-11-19 10:11:19.454 ERROR 1 --- [cTaskExecutor-1] n.g.v.s.rabbitmq.tester.MessageLogger    : Received message 00053408 - SKIPPED 1 MESSAGES.
2018-11-19 10:11:19.455 ERROR 1 --- [cTaskExecutor-1] n.g.v.s.rabbitmq.tester.MessageLogger    : Received message 00053395 - ARRIVED 13 MESSAGES TOO LATE.
2018-11-19 10:11:19.455 ERROR 1 --- [cTaskExecutor-1] n.g.v.s.rabbitmq.tester.MessageLogger    : Received message 00053410 - SKIPPED 14 MESSAGES.
2018-11-19 10:11:19.456 ERROR 1 --- [cTaskExecutor-1] n.g.v.s.rabbitmq.tester.MessageLogger    : Received message 00053397 - ARRIVED 13 MESSAGES TOO LATE.
2018-11-19 10:11:19.456 ERROR 1 --- [cTaskExecutor-1] n.g.v.s.rabbitmq.tester.MessageLogger    : Received message 00053399 - SKIPPED 1 MESSAGES.
2018-11-19 10:11:19.456 ERROR 1 --- [cTaskExecutor-1] n.g.v.s.rabbitmq.tester.MessageLogger    : Received message 00053412 - SKIPPED 12 MESSAGES.
2018-11-19 10:11:19.459 ERROR 1 --- [cTaskExecutor-1] n.g.v.s.rabbitmq.tester.MessageLogger    : Received message 00053414 - SKIPPED 1 MESSAGES.
2018-11-19 10:11:19.461 ERROR 1 --- [cTaskExecutor-1] n.g.v.s.rabbitmq.tester.MessageLogger    : Received message 00053401 - ARRIVED 13 MESSAGES TOO LATE.
```

## Code

componentQueue:
```
  @Bean public Queue componentQueue(@Value("${spring.application.name}") String applicationName) 
  {
    return new Queue(applicationName, true, false, false);
  }
```

MessageEmitter:
```
@Component
@EnableScheduling
public class MessageEmitter {

  private static final int TARGET_MPS = 200;

  private final RabbitMQTesterService rabbitMQTesterService;
  private final AtomicLong lastSentMessage = new AtomicLong(0);

  MessageEmitter(RabbitMQTesterService rabbitMQTesterService) {
    this.rabbitMQTesterService = rabbitMQTesterService;
  }

  @Scheduled(fixedDelay = 1000L / TARGET_MPS)
  public void sendMessage() {
    rabbitMQTesterService.sendMessage(lastSentMessage.incrementAndGet());
  }
}
```

rabbitListenerContainerFactory:
```
  @Bean
  public SimpleRabbitListenerContainerFactory rabbitListenerContainerFactory(ConnectionFactory connectionFactory) {
    SimpleRabbitListenerContainerFactory factory = new SimpleRabbitListenerContainerFactory();
    factory.setConnectionFactory(connectionFactory);
    factory.setConcurrentConsumers(1);
    factory.setMaxConcurrentConsumers(1);
    return factory;
  }
```

onMessageReceived:
```
  private final AtomicLong lastReceivedIndex = new AtomicLong(0L);

  public void onMessageReceived(long messageIndex) {
    long previousMessageIndex = lastReceivedIndex.getAndSet(messageIndex);

    if (messageIndex == previousMessageIndex + 1) {
      messageLogger.logSuccess(messageIndex);
    } else if (messageIndex == previousMessageIndex) {
      messageLogger.logDuplicate(messageIndex);
    } else if (messageIndex < previousMessageIndex) {
      messageLogger.logLate(messageIndex, previousMessageIndex - messageIndex);
    } else if (messageIndex > previousMessageIndex + 1) {
      messageLogger.logSkipped(messageIndex, messageIndex - previousMessageIndex - 1);
    }
  }
```

**Comment by lukebakken** [comment]
Hello and thanks for using RabbitMQ.

Please remember that RabbitMQ is maintained by a small team of engineers who do not have time to put together code to try and reproduce issues. The issue you describe "smells" like something specific to your application since we know that RabbitMQ correctly delivers messages, in publishing order, to consumers.

If you would like to continue discussing this issue, please post a message to the [`rabbitmq-users`](https://groups.google.com/forum/#!forum/rabbitmq-users) mailing list with code that we can run. Simply describing small parts of the code is not sufficient. Thank you.

**Comment by lukebakken** [comment]
I ran your code snippets past the maintainer of the RabbitMQ Java client (@acogoluegnes), and the snippets appear to be correct.

One thing you can do to investigate is to do the following:

* Run your test scenario
* While it is running, run the following command on your RabbitMQ node:

```
rabbitmqctl list_consumers
```

You should only see one consumer registered, like this:

```
Listing consumers on vhost / ...
queue_name      channel_pid     consumer_tag    ack_required    prefetch_count  arguments
amq.gen-iB4VTjl9u469x1ggQHLw2Q  <rabbit@shostakovich.2.624.0>   amq.ctag-uFeAs5hOYf55vjv3irE1-Q true    0       []
```

If you see more than one consumer, it mean that the `@Bean` configuration you have provided is not being applied correctly. This could be an issue in your application (likely) or a bug in Spring Boot (less likely). If you post a complete set of code to `rabbitmq-users`, I will make sure that a member of the Spring team sees it.

**Comment by Oduig** [comment]
I agree that it sounds like something outside RabbitMQ, otherwise we could expect more issues to be reported. At the same time it's quite some work to set up a runnable example, so it helps to know the effort will not be fruitless. Thanks! I'll put it up on GitHub and contact the user group.

P.S. `list_consumers` returns a single consumer, just like your example

**Comment by lukebakken** [comment]
> P.S. list_consumers returns a single consumer, just like your example

I'm surprised by that result, so thanks for letting us know. I will keep an eye out on the mailing list for your code example.

**Comment by Oduig** [comment]
Hi, just an FYI for those reading this. I thought I lost my code after changing laptops so I never got around to posting it to the user group. Turns out we still had it somewhere.

I extracted all the code into a minimal example and ran it against a separate K8s cluster. Whatever the problem was before, it seems to have gone away and I can no longer reproduce it. I assume it's some difference in configuration or cluster setup. Let you know if we run into it again.
